exec/java-exec/src/test/java/org/apache/drill/TestUnionAll.java

```diff
@@ -1013,4 +1019,97 @@ public void testUnionAllInWith() throws Exception {
         .build()
         .run();
   }
+
+  @Test // DRILL-4147 // base case
+  public void testDrill4147_1() throws Exception {
+    final String l = FileUtils.getResourceAsFile("/multilevel/parquet/1994").toURI().toString();
+    final String r = FileUtils.getResourceAsFile("/multilevel/parquet/1995").toURI().toString();
+
+    final String query = String.format("SELECT o_custkey FROM dfs_test.`%s` \n" +
+        "Union All SELECT o_custkey FROM dfs_test.`%s`", l, r);
+
+    // Validate the plan
+    final String[] expectedPlan = {"UnionExchange.*\n",
+        ".*Project.*\n" +
+        ".*UnionAll"};
+    final String[] excludedPlan = {};
+
+    test(sliceTargetSmall);
+    PlanTestBase.testPlanMatchingPatterns(query, expectedPlan, excludedPlan);
+
+    try {
+      testBuilder()
+        .optionSettingQueriesForTestQuery(sliceTargetSmall)
+        .optionSettingQueriesForBaseline(sliceTargetDefault)
+        .unOrdered()
+        .sqlQuery(query)
+        .sqlBaselineQuery(query)
+        .build()
+        .run();
+    } finally {
+      test(sliceTargetDefault);
+    }
+  }
+
+  @Test // DRILL-4147  // group-by on top of union-all
+  public void testDrill4147_2() throws Exception {
+    final String l = FileUtils.getResourceAsFile("/multilevel/parquet/1994").toURI().toString();
+    final String r = FileUtils.getResourceAsFile("/multilevel/parquet/1995").toURI().toString();
+
+    final String query = String.format("Select o_custkey, count(*) as cnt from \n" +
+        " (SELECT o_custkey FROM dfs_test.`%s` \n" +
+        "Union All SELECT o_custkey FROM dfs_test.`%s`) \n" +
+        "group by o_custkey", l, r);
+
+    // Validate the plan
+    final String[] expectedPlan = {"(?s)UnionExchange.*HashAgg.*HashToRandomExchange.*UnionAll.*"};
+    final String[] excludedPlan = {};
+
+    test(sliceTargetSmall);
```

Comment:
Yes, I will move this and other such calls within the try block.  Thanks for pointing out.